smack-core/src/main/java/org/jivesoftware/smack/remoteEndpoint/HttpLookUp.java

```diff
@@ -0,0 +1,197 @@
+
+    String obtainedXml = "";
+
+    /**
+     * Constructs a HTTP connection with the server and retrieves complete host metadata.
+     *
+     * @param host host that needs to be connected
+     * @param usingHttps true if secure HTTP being used, false otherwise
+     * @throws IOException exception due to input/output operations
+     */
+    public HttpLookUp(String host, boolean usingHttps) throws IOException {
+        final String metadataUrl = (usingHttps ? "https://" : "http://") + host + "/.well-known/host-meta";
+        httpConnection = getHttpConnection(metadataUrl);
+        content_length = httpConnection.getContentLength();
+
+        InputStream inputStream = httpConnection.getInputStream();
+        int remainingBytes = content_length;
+
+        // Keep reading from the input stream until read completely
+        while (remainingBytes > 0) {
+            byte[] byteBuffer = new byte[BUFFER_SIZE];
+            int bytes_read = inputStream.read(byteBuffer);
+            obtainedXml += new String(byteBuffer, 0, bytes_read, Charset.defaultCharset());
+            remainingBytes = remainingBytes - bytes_read;
+        }
+    }
+
+    private static HttpURLConnection getHttpConnection(String endpoint) throws IOException {
+        final URL putUrl = new URL(endpoint);
+        HttpURLConnection httpURLConnection = (HttpURLConnection) putUrl.openConnection();
+        httpURLConnection.setRequestMethod("GET");
+        httpURLConnection.setUseCaches(false);
+        httpURLConnection.setDoOutput(false);
+        httpURLConnection.setRequestProperty("Content-Type", "xml");
+        return httpURLConnection;
+    }
+
+    /**
+     * Get remote endpoints from the web host metadata fetched from the server.
+     *
+     * @param relAttribute {@link RelAttribute} for the type of connection
+     * @return List of {@link RemoteConnectionEndpoint}s
+     * @throws IOException exception due to input/output operations
+     * @throws XmlPullParserException exception encountered during XML parsing
+     * @throws URISyntaxException exception to indicate that a string could not be parsed as a URI reference
+     */
+    public List<RemoteConnectionEndpoint> getEndpoints(RelAttribute relAttribute)
+                    throws IOException, XmlPullParserException, URISyntaxException {
+        return getEndpoints(relAttribute.attribute);
+    }
+
+    /**
+     * Get remote endpoints from the web host metadata fetched from the server.
+     *
+     * @param relAttribute relation attribute for the type of connection
+     * @return List of {@link RemoteConnectionEndpoint}s
+     * @throws IOException exception due to input/output operations
+     * @throws XmlPullParserException exception encountered during XML parsing
+     * @throws URISyntaxException exception to indicate that a string could not be parsed as a URI reference
+     */
+    public List<RemoteConnectionEndpoint> getEndpoints(String relAttribute)
```

Comment:
Can you see what might be non-optimal and will happen if the user does the following:
```
if (httpLookUp.getEndpoints("urn:xmpp:alt-connections:websocket").size() > 0) {
  for (RemoteConnectionEndpoint e : httpLookUp.getEndpoints("urn:xmpp:alt-connections:websocket")) {
    ...
  }
}
```

Reply:
Also, while I like that this method is generic enough to enable use of other, currently unknown endpoint namespaces, I would also like to see some explicitly named methods like `getWebsocketEndpoints()`.

Reply:
While `relAttribute` is probably the correct technical term, you may consider naming the argument something else, eg. `protocolNamespace` so that it is immediatelly clear what the method does.

Reply:
Also, while I like that this method is generic enough to enable use of other, currently unknown endpoint namespaces, I would also like to see some explicitly named methods like `getWebsocketEndpoints()`.
While `relAttribute` is probably the correct technical term, you may consider naming the argument something else, eg. `protocolNamespace` so that it is immediately clear what the method does.

Reply:
A [hint](https://github.com/igniterealtime/Smack/blob/master/smack-core/src/main/java/org/jivesoftware/smack/util/MultiMap.java) for what you might use when parsing out the endpoints :)

Reply:
> Can you see what might be non-optimal and will happen if the user does the following:
> 
> ```
> if (httpLookUp.getEndpoints("urn:xmpp:alt-connections:websocket").size() > 0) {
>   for (RemoteConnectionEndpoint e : httpLookUp.getEndpoints("urn:xmpp:alt-connections:websocket")) {
>     ...
>   }
> }
> ```

We end up parsing the same obtainedXml twice :)